### ISSUE TITLE
Usage example of train_mnist?

### ISSUE THREAD

**Issue by berndporr** [artyom-beilis/dlprimitives#37]
So I can run train_mnist:
```
~/sandbox/dlprimitives$ ./train_mnist 
Usage device net.json mnist_img mnist_lbl
```
Is there a complete workable usage example with example parameters? No idea what these mean but it looks exactly what I'm looking for to hack it (planning to do realtime learning as done in my deep Neuronal Filter: https://github.com/berndporr/deepNeuronalFilter). So basically inference & training at every time-step.

**Comment by artyom-beilis** [comment]
From build directory:

    ./train_mnist 0:0 ../tools/nets/mnist_train.json ../data/MNIST/raw/train-images-idx3-ubyte ../data/MNIST/raw/train-labels-idx1-ubyte

Where `0:0` is platform:device next the net for training 3rd and 4th are MNIST raw data as downloaded and gunzipped.

However I strongly recommend to take a look https://github.com/artyom-beilis/pytorch_dlprim - pytorch backend based on dlprimitives - it is way more full featured training system that supports much more operators and features.

Dlprimitives as training/inference framework support very basic nets (something like Caffe) it works in static graph and operator set is very-very limited.

If future I plan to invest in pytorch backend and ONNX inference support.

**Comment by berndporr** [comment]
The py torch wrapper has far too much overhead. What I ideally need is forward and backward pass every 20us. So needs to be C++ but as much as possible in the kernels. Just now I just need a fully connected encoder with 3 layers and tanh. That works pretty well with CUDA but I get max 2kHz out of it. So no python for sure. For that reason I love your library. Even that it's a bit limited it far far more complex than my deep neuronal filter or CLDL. I'm rather after super low training times in the region of ms or us.

**Comment by berndporr** [comment]
Thanks. Works:
```
./train_mnist 0:0 tools/nets/mnist_train.json /tmp/train-images-idx3-ubyte /tmp/train-labels-idx1-ubyte 
Using: Intel(R) HD Graphics 5500 [0x1616] on Intel(R) OpenCL HD Graphics
Start training
Epoch/iter 0/100 loss=0.512333 acc=0.84517
```

**Comment by artyom-beilis** [comment]
> The py torch wrapper has far too much overhead

There is C++ API for pytorch, but if you are only running a small net it indeed can be easier to use dlprimitives

**Comment by berndporr** [comment]
Wondering how to inject the loss directly here:
https://github.com/artyom-beilis/dlprimitives/blob/master/examples/cpp/train_mnist.cpp#L118
As I said above I calc a different loss which is not based on labels vs output. I've gone through the code for two last two hours but couldn't find it. Also would be happy to add a function to add that properly. As I said I'm doing realtime noise cancellation and there the error (loss) is the difference between the network output (just one neuron, i.e. encoder) and a noisy signal:
![Adaptive-noise-cancelling-using-LMS-filter](https://github.com/user-attachments/assets/3274fbec-2c5b-4a20-a853-c147b926c68c)
thunk of the adaptive box of the deep network. At every audio sample it's propagated through the network. Then an error is calculated (i.e. loss) and then the weights are changed. I guess that would be your SGD solver. So it's all there but I have no idea where in your code I could set the error/loss.
